### PR TITLE
[Workflow] Fixed graphviz dumper for state machine

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #20497 
| License       | MIT
| Doc PR        | ~

---

Before:
before
After:
after

---

Script:
```php
<?php

require __DIR__.'/vendor/autoload.php';

use Symfony\Component\Workflow\Definition;
use Symfony\Component\Workflow\Dumper\StateMachineGraphvizDumper;
use Symfony\Component\Workflow\Marking;
use Symfony\Component\Workflow\StateMachine;
use Symfony\Component\Workflow\Transition;

$places = array('a', 'b', 'c', 'd');
$transitions[] = new Transition('t1', 'a', 'b');
$transitions[] = new Transition('t1', 'd', 'b');
$transitions[] = new Transition('t2', 'b', 'c');
$transitions[] = new Transition('t3', 'b', 'd');
$definition = new Definition($places, $transitions);

$marking = new Marking(['d' => 1]);

echo (new StateMachineGraphvizDumper())->dump($definition, $marking);
```

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php [hunk 1]
```diff
@@ -60,13 +61,14 @@ protected function execute(InputInterface $input, OutputInterface $output)
         $serviceId = $input->getArgument('name');
         if ($container->has('workflow.'.$serviceId)) {
             $workflow = $container->get('workflow.'.$serviceId);
+            $dumper = new GraphvizDumper();
         } elseif ($container->has('state_machine.'.$serviceId)) {
             $workflow = $container->get('state_machine.'.$serviceId);
+            $dumper = new StateMachineGraphvizDumper();
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php]: should we have a way to dump the state machine in workflow mode ? It is a subset after all.

But anyway, I agree it should not be the default dumping for state machines.

Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php]: this of course means that it should be fixed to work properly, but I know how to do this (stop assuming that transition names are unique identifiers of the transition, as we stopped doing this.

File: src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php [hunk 2]
```diff
@@ -93,37 +93,29 @@ private function findTransitions(Definition $definition)
...
-                $this->addAttributes($place['attributes'])
-            );
+            $code .= sprintf("  transition_%s [label=\"%s\", shape=box%s];\n", $this->dotize($place['name']), $place['name'], $this->addAttributes($place['attributes']));
         }
 
         return $code;
     }
 
-    private function findEdges(Definition $definition)
+    protected function findEdges(Definition $definition)
```
Comment (stof) [inline src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php]: IMO, this should not be protected extension points. This seems like the wrong way to reuse part of the GraphvizDumper (fixing the dumping of transitions to stop assuming unique names might require refactoring private methods, but making them protected forbids this.

File: src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php (hunk 1 above)
Comment (lyrixx) [inline src/Symfony/Bundle/FrameworkBundle/Command/WorkflowDumpCommand.php]: > should we have a way to dump the state machine in workflow mode ? 

For now, it's not really readable to dump a state machine with the workflow dumper. So IMHO, it's not useful to add a way to do it.

File: src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php (hunk 2 above)
Comment (lyrixx) [inline src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php]: I don't really understand what you mean. Could you elaborate?

Comment (lyrixx) [inline src/Symfony/Component/Workflow/Dumper/GraphvizDumper.php]: I added `@internal` everywhere.

**Comment by nicolas-grekas** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @lyrixx.

**Comment by chalasr** [comment]
Unless I missed something, this should have been merged in 3.2 instead

**Comment by fabpot** [comment]
Indeed, thanks for noticing.

**Comment by fabpot** [comment]
Thank you @lyrixx.

**Pull request merged**